### PR TITLE
crimson/osd/snaptrim_event: Handle missing clone obc case

### PR DESCRIPTION
**Pull request by Matan-B**
```
The case where resolve_oid is not able to return the clone oid (in order
to load the clone obc) should result in an SNAPTRIM_ERROR and stop the loop
iteration. Without taking this case into account an infinite loop of snap
trim events will be created.
A comment was also added to reference classic's approach for this case.
```

Fixes: https://tracker.ceph.com/issues/59231

Blocked by: https://github.com/ceph/ceph/pull/51180

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by athanatos — APPROVED** [review]

File: src/crimson/osd/osd_operations/snaptrim_event.cc [hunk 1]
```diff
@@ -148,6 +149,11 @@ SnapTrimEvent::with_pg(
             pg,
             object,
             snapid);
+          if (pg->get_peering_state().state_test(PG_STATE_SNAPTRIM_ERROR)) {
```
Comment (rzarzynski) [inline src/crimson/osd/osd_operations/snaptrim_event.cc:152]: As far as I recall the original idea was to minimize the # communication channels and keep the flag management in the client of `SnapTrimEvent`. It already does:

```cpp
        return shard_services.start_operation<SnapTrimEvent>(
          this,
          snap_mapper,
          to_trim,
          needs_pause
        ).second.handle_error(
          crimson::ct_error::enoent::handle([this] {
            logger().error("{}: ENOENT saw, trimming stopped", *this);
            peering_state.state_set(PG_STATE_SNAPTRIM_ERROR);
            publish_stats_to_osd();
            return seastar::make_ready_future<seastar::stop_iteration>(
              seastar::stop_iteration::yes);
          }), 
```

Here `PG_STATE_SNAPTRIM_ERROR` is being used as a side-channel to the error-in-return-future.
Why not let `SnapTrimObjSubEvent` simply returned a failed `remove_or_update_iertr::future<>`?

File: src/crimson/osd/osd_operations/snaptrim_event.cc [hunk 2]
```diff
@@ -524,9 +530,17 @@ SnapTrimObjSubEvent::with_pg(
...
-      return seastar::now();
-    }));
+    }).handle_error_interruptible(
+      // See how classic handles this case at the beginning of
+      // PrimaryLogPG::trim_object,
+      crimson::ct_error::enoent::handle([this](auto &) {
+        logger().error("{}: saw ENOENT while loading obc,"
+                       "setting pg state to SNAPTRIM_ERROR", *this);
+        pg->get_peering_state().state_set(PG_STATE_SNAPTRIM_ERROR);
+        return seastar::now();
```
Comment (rzarzynski) [inline src/crimson/osd/osd_operations/snaptrim_event.cc:540]: That's what the previous comment is about. Conceptually returning `now()` informs that an error has been _handled_ but here we just repack / redirect it to a new communication channel. Communication channe that strongly glues parts of loops of `SnapTrimObjSubEvent` and `SnapTrimEvent` together.

**Review by rzarzynski — CHANGES_REQUESTED** [review]
First and foremost big thanks for nailing this down; good job, Matan!

In the comments I'm grumbling a bit about handling the `enoent` error. I wonder whether we could simplify the fix while preserving original intentions behind structuring the code.

File: src/crimson/osd/osd_operations/snaptrim_event.cc [hunk 3]
```diff
@@ -524,9 +530,17 @@ SnapTrimObjSubEvent::with_pg(
           });
         });
       });
-    }).handle_error_interruptible(PG::load_obc_ertr::all_same_way([] {
-      return seastar::now();
```
Comment (rzarzynski) [inline src/crimson/osd/osd_operations/snaptrim_event.cc:528]: Could we just propagate `enoent` and assert on everything else?

File: src/crimson/osd/osd_operations/snaptrim_event.cc (hunk 1 above)
Comment (Matan-B) [reply src/crimson/osd/osd_operations/snaptrim_event.cc]: The error is now passed_further to be handled by the  the client of `SnapTrimEvent`.

**Review by rzarzynski — APPROVED** [review]

**Comment by Matan-B** [comment]
Merging based on:
https://pulpito.ceph.com/matan-2023-04-25_13:23:05-crimson-rados-wip-matanb-crimson-only-testing-25.4-distro-crimson-smithi/

**Pull request merged**